File: sections/09_greedy_exchange/problems/b_rescue_boats/solution.py
"""Reference solution for B. Rescue Boats."""

import sys


def solve_case(weights: list[int], limit: int) -> int:
    weights.sort()
    left = 0
    right = len(weights) - 1
    boats = 0
    while left <= right:
        if weights[left] + weights[right] <= limit:
            left += 1
        right -= 1
        boats += 1
    return boats


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        limit = data[idx + 1]
        idx += 2
        weights = data[idx:idx + n]
        idx += n
        out.append(str(solve_case(weights, limit)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/09_greedy_exchange/problems/b_rescue_boats/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int solve_case(vector<int> weights, int limit) {
    sort(weights.begin(), weights.end());
    int left = 0;
    int right = (int)weights.size() - 1;
    int boats = 0;
    while (left <= right) {
        if (weights[left] + weights[right] <= limit) {
            ++left;
        }
        --right;
        ++boats;
    }
    return boats;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n, limit;
        cin >> n >> limit;
        vector<int> weights(n);
        for (int &w : weights) cin >> w;
        cout << solve_case(weights, limit) << '\n';
    }

    return 0;
}

